### DISCUSSION TITLE
How to use the For Loop tool in Sirius?

### DISCUSSION THREAD

**Discussion by t-chiang** [eclipse-sirius/sirius-desktop#477] [Q&A]
Hello all,

I'm trying to use the For Loop operation to iterate through objects in a container to set their attributes. I have the loop currently set up as the following:

![image](https://github.com/user-attachments/assets/ec93f16d-078c-44e8-9ede-3afa829bf0e2)

For the Set operations I then use the 'children' iterator to access the objects within my container to set their attributes:
![image](https://github.com/user-attachments/assets/bcab2317-abf6-4062-9efe-b15c9c58a556)

Unfortunately this isn't working and I'm unsure why and what the correct way to do this operation is. Any help advice?

Side note: is there a way to use AQL to print to console to help with debugging? Or any programming tips in AQL to see the results of a query and think about what the next step might be?

**Comment by lredor** [comment]
Hi T-Chiang,
Here is a sample of an action that renames all the attributes of an EClass : 
[ForSample.zip](https://github.com/user-attachments/files/17329283/ForSample.zip)
![2024-10-10_16h15_38](https://github.com/user-attachments/assets/08d8e7fe-561d-40dd-b7c7-a9efe954321f)
![2024-10-10_16h00_52](https://github.com/user-attachments/assets/006751fb-f269-42a2-a112-958ab88a4ecd)
![2024-10-10_16h01_03](https://github.com/user-attachments/assets/83ac9338-a7d6-45f4-b5b3-e839c78647c6)

**Reply by t-chiang** [reply]
Hi Laurent,

This is almost what I'm looking for, instead of the attribute within an object, I want to modify the attributes of the objects within my container:
![image](https://github.com/user-attachments/assets/dc0e0e3a-eaee-4ad4-9627-a1229d63e2f7)

Following your example I've removed the 'Let' portion of my operation so I now have:
![image](https://github.com/user-attachments/assets/33439866-0ed5-4926-b84d-5adeb3164ce8)

Edit:
I have since narrowed down the problem not to the for-loop, but the AQL command that returns the container from the iterator.
![image](https://github.com/user-attachments/assets/3589bb96-1915-4bc3-b97f-6894278da82c)
In the image above, my set command returns null for some reason. The same command without an iterator works perfectly fine in another menu option so I'm confused about what is preventing me from accessing the features of my container from the for-loop. The only feature I can access the name attribute and nothing else.

**Reply by lredor** [reply]
Another sample that uses the package parent name in the value expression (using OD 11.8.0 and so Sirius 7.1.0): 
[ForSample2.zip](https://github.com/user-attachments/files/17341275/ForSample2.zip).

In your case, I'm not sure about:
- On which element is the action launched?
- What is children ? What is the expression of the `For` operation?

**Reply by t-chiang** [reply]
This is a sample of my metamodel:
![image](https://github.com/user-attachments/assets/06733ea9-3310-4e7e-befe-7b20cb1d354a)
What I'm trying to do with this for loop is copy all the references from the RequirementContainer instance to all of its children (semantic elements of type 'RequirementType'). For some reason, when I try to access the semantic elements that are connected to the container and copy them to the children the AQL queries fail and return NULL. So any command I do within the context of the container like "aql:self.traceto.name" returns NULL. I know that self isn't NULL because "aql:self.name" returns the name of the container that I'm using the operation on. It's weird because this only happens on operations that use the precondition "aql:self.oclIsTypeOf(rmdl::RequirementContainer)".

**Reply by lredor** [reply]
It's complicated to help you further without starting with a concrete case. I therefore suggest that you try to reproduce your problematic case on the Ecore metamodel by providing a reproduction project. I can then try to look in a little more detail.

**Comment by lredor** [comment]
And for the question `is there a way to use AQL to print to console to help with debugging`, there is no real debug feature of AQL expression in the VSM. You can create a Java service "debug(EObject)" that returns the object itself and displays some "sysout". You have an example of this approach [here](https://github.com/ObeoNetwork/InformationSystem/blob/master/commons/utils/plugins/org.obeonetwork.utils.common.ui/src/org/obeonetwork/utils/common/ui/services/DebugServices.java).
